File: app/src/main/java/org/rivchain/cuplink/PacketReader.java
package org.rivchain.cuplink;

import java.io.IOException;
import java.io.InputStream;
import java.net.Socket;


/* Read the message size from the header and return the message of the correct size */
class PacketReader {
    final InputStream is;
    final byte[] buffer;
    int pos;

    PacketReader(Socket socket) throws IOException {
        this.is = socket.getInputStream();
        this.buffer = new byte[16000];
        this.pos = 0;
    }

    private static int readMessageHeader(byte[] packet) {
        return ((packet[0] & 0xFF) << 24)
                | ((packet[1] & 0xFF) << 16)
                | ((packet[2] & 0xFF) << 8)
                | ((packet[3] & 0xFF) << 0);
    }

    byte[] readMessage() throws IOException {
        while (true) {
            int read = is.read(buffer, pos, buffer.length - pos);
            if (read < 0) {
                break;
            }

            pos += read;

            if (pos < 4) {
                continue;
            }

            int len = readMessageHeader(buffer);
            if (len < 0 || len > buffer.length) {
                break;
            }

            // not enough data
            if (pos < (4 + len)) {
                continue;
            }

            byte[] request = new byte[len];
            System.arraycopy(buffer, 4, request, 0, request.length);

            // move data of next messgae to the front of the buffer
            System.arraycopy(buffer, 4 + len, buffer, 0, pos - request.length);
            pos = 0;
            return request;
        }

        return null;
    }

    private void log(String s) {
        Log.d(this, s);
    }
}

File: app/src/main/java/d/d/meshenger/PacketReader.kt
package d.d.meshenger

import d.d.meshenger.Log.d
import kotlin.Throws
import d.d.meshenger.PacketReader
import java.io.IOException
import java.io.InputStream
import java.net.Socket
import kotlin.experimental.and

/* Read the message size from the header and return the message of the correct size */
internal class PacketReader(socket: Socket) {
    val `is`: InputStream
    val buffer: ByteArray
    var pos: Int
    @Throws(IOException::class)
    fun readMessage(): ByteArray? {
        while (true) {
            val read = `is`.read(buffer, pos, buffer.size - pos)
            if (read < 0) {
                break
            }
            pos += read
            if (pos < 4) {
                continue
            }
            val len = readMessageHeader(buffer)
            if (len < 0 || len > buffer.size) {
                break
            }

            // not enough data
            if (pos < 4 + len) {
                continue
            }
            val request = ByteArray(len)
            System.arraycopy(buffer, 4, request, 0, request.size)

            // move data of next messgae to the front of the buffer
            System.arraycopy(buffer, 4 + len, buffer, 0, pos - request.size)
            pos = 0
            return request
        }
        return null
    }

    private fun log(s: String) {
        d(this, s)
    }

    companion object {
        private fun readMessageHeader(packet: ByteArray): Int {
            return (packet[0].toInt() and 0xFF shl 24
                    or (packet[1].toInt() and 0xFF shl 16)
                    or (packet[2].toInt() and 0xFF shl 8)
                    or (packet[3].toInt() and 0xFF shl 0))
        }
    }

    init {
        `is` = socket.getInputStream()
        buffer = ByteArray(16000)
        pos = 0
    }
}